pytorch_lightning/trainer/training_loop.py

```diff
@@ -561,17 +561,18 @@ def run_training_epoch(self):
 
         should_check_val = self.should_check_val_fx(batch_idx, is_last_batch, on_epoch=True)
         if should_check_val:
-            self.trainer.run_evaluation(on_epoch=True)
+            self.trainer.run_evaluation()
```

Comment:
okay I see the problem here. what we can do there is have a flag (bool) or something that will indicate whether val_loop is called or not inside the training loop (val_check_interval < 1.), and if it's true then we can just update the lr before the val loop outside training loop is triggered to solve both the conditions here. 

```python
if should_val_check or val_loop_called:
    self.optimizer_connector.update_learning_rates(interval='epoch')

if should_val_check:
    self.trainer.run_evaluation()
```

will this work?

Reply:
learning rates needs to be updated after `run_evaluation` since some schedulers like `
ReduceLROnPlateau` can be condition on specific metrics calculated in the validation loop

Reply:
then the checkpoint won't be getting the correct state since `on_validation_end` for checkpoint callback is called inside validation_loop. I am assuming that before on_validation_end call in the val loop, everything logged inside `LightningModule` hooks or log-supported callback hooks must be aggregated correctly. If they aren't then that would be a problem with other callbacks like `ModelCheckpoint`, `EarlyStopping` too right?

@tchaton can you confirm whether everything logged is aggregated before [this call](https://github.com/PyTorchLightning/pytorch-lightning/blob/0b271474e5cc35a40009559d6c4488ab3b02723e/pytorch_lightning/trainer/trainer.py#L758) or not?

Reply:
hmm.. I think 
```
if val_loop_called:
    self.optimizer_connector.update_learning_rates(interval='epoch')
if should_val_check:
    self.trainer.run_evaluation(on_epoch=True)
```
should be enough and rest as it is. I hope didn't miss any edge case this time.

Although there is still one more edge case here. If user sets `val_check_interval=0.3` or something then metric used might not be the correct one that will be used to make decision inside `ReductLRonPleateau`, but I'd say that is more of the logic/implementation error from user's side.

Reply:
alright I try to revert back, and see if it fixes the problem

Reply:
oh sorry! missed one more edge case here :sweat_smile: 

```
if val_loop_called and not should_check_val:
```
to handle cases when val_check_interval=0.1, 0.2, 0.25, 0.5

thinking for more edges cases here to avoid any possible failures in future.

Reply:
> thinking for more edges cases here to avoid any possible failures in future.

Can we write tests for all these corner cases? Better safe than sorry

Reply:
added some tests. found a bug too.

Reply:
Is this ready-to-go then?

Reply:
Need to add a similar strong test for reducelronplateau scheduler.